### ISSUE TITLE
Export Data to GeoKey script

### ISSUE THREAD

**Issue by spMohanty** [geotagx/geotagx-theme#66]

**Comment by spMohanty** [comment]
As described in http://geokey.org.uk/docs/web/contribution-add.html  
Every GeoKey contribution needs to be a valid GeoJson (hence every contribution has to have a geolocation associated with it)
In `summary` of all task_runs for a particular project, we now have all the data aggregated in the format below :

``` javascript
{  
   "http://images.indianexpress.com/2013/06/b_id_395026_yamuna-2.jpg":{  
      "shelter2013::dwellingTemp":{  
         "GEOTAGX_UNANSWERED":1,
         "No":1
      },
      "shelter2013::dwellingFlood":{  
         "Inundated":2
      },
      "shelter2013::waterLargeFast":{  
         "Unknown":1
      },
      "shelter2013::dwellingMaterial":{  
         "GEOTAGX_UNANSWERED":1,
         "Concrete, Iron":1
      },
      "shelter2013::waterLarge":{  
         "No":1
      },
      "shelter2013::dwellingType":{  
         "Public":1,
         "Private":1
      },
      "shelter2013::waterPeopleActivities":{  
         "Done":1
      },
      "shelter2013::GEOTAGX_TOTAL":2,
      "shelter2013::dwellingConstruction":{  
         "GEOTAGX_UNANSWERED":1,
         "Pucca":1
      },
      "shelter2013::dwelling":{  
         "Yes":2
      }
   },
   "http://images.indianexpress.com/2013/06/b_id_395243_vehicles.jpg":{  
      "shelter2013::dwellingTemp":{  
         "GEOTAGX_UNANSWERED":1
      },
      "shelter2013::dwellingFlood":{  
         "GEOTAGX_UNANSWERED":1
      },
      "shelter2013::waterLargeFast":{  

      },
      "shelter2013::dwellingMaterial":{  
         "GEOTAGX_UNANSWERED":1
      },
      "shelter2013::waterLarge":{  

      },
      "shelter2013::dwellingType":{  
         "GEOTAGX_UNANSWERED":1
      },
      "shelter2013::waterPeopleActivities":{  

      },
      "shelter2013::GEOTAGX_TOTAL":2,
      "shelter2013::dwellingConstruction":{  
         "GEOTAGX_UNANSWERED":1
      },
      "shelter2013::dwelling":{  
         "No":2
      }
   },
   "http://s4.firstpost.in/wp-content/uploads/2013/06/5DelhiFloods_AP.jpg":{  
      "shelter2013::dwellingTemp":{  
         "GEOTAGX_UNANSWERED":1
      },
      "shelter2013::dwellingFlood":{  
         "Inundated":1
      },
      "shelter2013::waterLargeFast":{  
         "Unknown":1
      },
      "shelter2013::dwellingMaterial":{  
         "GEOTAGX_UNANSWERED":1
      },
      "shelter2013::waterLarge":{  
         "No":1
      },
      "shelter2013::dwellingType":{  
         "Private":1
      },
      "shelter2013::waterPeopleActivities":{  
         "Done":1
      },
      "shelter2013::GEOTAGX_TOTAL":1,
      "shelter2013::dwellingConstruction":{  
         "GEOTAGX_UNANSWERED":1
      },
      "shelter2013::dwelling":{  
         "Yes":1
      }
   }
}
```

In the above data structure the image is the `key` of every entry (as is in the design of geotagx), while geokey expects that the geolocation of the summary be the key.

Now that we **cannot** directly upload the summary, we will have to individually merge the summaries from all projects in the same geotagx-categories, then check if there is one valid geolocation field, and then send the whole data to GeoKey as a contribution in GeoJson format, with type as `Polygon`

In terms of GeoKey we will have one geokey-category of the same `geotagx-summary` with one `TextField` by the name `info` which will hold the JSON summary for the whole image.

Everytime we **sync** the geotagx and geokey data, a new geokey-contribution will be added for every image in geotagx.  Please note that, by this approach the contributions in geokey will not be independent records, but instead be incremental records for the same image as more and more TaskRuns are recorded.

**Comment by spMohanty** [comment]
Linked to https://github.com/geotagx/geotagx-theme/issues/62

**Comment by oliverroick** [comment]
As I understood, you will get several geotags for an image. You could (based on a sufficient number of geotags) calculate a spatial centroid and use that for the location for an images or you calculate the convex hull of all points and use the resulting polygon as the geometry. 

The POST could like:

```

{
  "type": "Feature",
  "properties": {
    "info": "Summary",
    "image-link": "http://s4.firstpost.in/wp-content/uploads/2013/06/5DelhiFloods_AP.jpg"
  },
  "geometry": {
    "type": "Polygon",
    "coordinates": [
      [
        [
          -0.12153625488281249,
          51.5540206177
        ],
        [
          -0.1242828369140625,
          51.536085601784755
        ],
        [
          -0.0858306884765625,
          51.53864817973767
        ],
        [
          -0.10505676269531249,
          51.551885867448604
        ],
        [
          -0.12153625488281249,
          51.5540206177
        ]
      ]
    ]
  }
}
```

Once you post that to GeoKey you will get a response of the feature created with an ID of the feature in GeoKey. You could store that ID with the corresponding task and use it to update the object in GeoKey, when you have new data. See: http://geokey.org.uk/docs/web/contribution-update.html

**Comment by spMohanty** [comment]
Hi @Oliverroick,
Can you check if this is what you had in mind : http://geotagx.org/geotagx/export/category/emergencyshelterassessmentinthemiddleeast/GeoJSON ?

**Comment by oliverroick** [comment]
Almost:
1. GeoKey does not support bulk upload, POSTing a `FeatureCollection` will return an error. You have to upload all features individually.
2. You can upload complex objects as properties it won't throw an error; but GeoKey can't validate these as it supports only simple data types (text, numbers, dates, etc.). Also the client that visualises the results, must know about the structure of complex objects as GeoKey can't provide meta information (Field names, types, etc.) for these, but I think thats not necessary for GeoTag-X.

**Comment by oliverroick** [comment]
I've built a quick prototype for a map viewer. This is what it looks like:

![screen shot 2015-07-14 at 13 01 59](https://cloud.githubusercontent.com/assets/159510/8672674/924a7574-2a28-11e5-80b5-3860d16d886d.png)

I created an example project that you can use for testing:
- http://play.geokey.org.uk/geotagx/86/ That's the map
- Use this URL to post your results: http://play.geokey.org.uk//api/geotagx/86/import/

Here's the repo + some docs: https://github.com/ExCiteS/geokey-geotagx

/cc @elliec

**Comment by spMohanty** [comment]
Hi olivier,
On trying to access the [link](http://play.geokey.org.uk/geotagx/86/) I see this ?? :

![screen shot 2015-07-18 at 6 11 58 pm](https://cloud.githubusercontent.com/assets/1581312/8762432/e8c1aa82-2d78-11e5-9db7-0c72f32ceb1a.png)

**A Quick Update : I just now changed the projection in our GeoJSON export feature from WebMercator  to WGS84, I think that might fix this issue now. Will check and get back to you soon.**

**Comment by oliverroick** [comment]
Yeah, you should change the projection to WGS84. 

Also, the error might be because the list of features is empty. Have you uploaded any data yet?

**Comment by elliec** [comment]
@spMohanty to check how it works with new projection and then if ok write cron job

**Comment by spMohanty** [comment]
Putting here the contents of reply over email for future reference :

---

@oliverroick : I get a TemplateNotFound error when I access the link : http://play.geokey.org.uk/api/geotagx/87/import/

We had changed the projection of the coordinates WGS84, and we also developed a GeoJSON visualiser for the results specific to the GeoTag-X which works by just using the GeoJSON export. 
But we can still push the summary of the results in GeoJSON to GeoKey.
Let me know if a cron job pushing the results to the url above can work after the issue with the missing template is resolved. Then I can set one up quickly.

---

**Comment by oliverroick** [comment]
Have you accessed the endpoint from a browser? You usually get the error with Django, when you access an endpoint that requires application/json as the content-type. Also HTTP GET is not allow on that endpoint.

 Try POSTing data to the endpoint using content-type application/json, that should do.

**Comment by spMohanty** [comment]
@oliverroick : Sorry I posted the wrong link. I meant the map : http://play.geokey.org.uk/geotagx/86/

**Comment by spMohanty** [comment]
@oliverroick : My bad. its working now. Though I got this a few minutes ago : ![screenshot](http://i.imgur.com/IDTSp2B.png)

**Comment by oliverroick** [comment]
@spMohanty That's the wrong link. You have used http://play.geokey.org.uk/ **api** /geotagx/86/ instead of http://play.geokey.org.uk/geotagx/86/

**Comment by spMohanty** [comment]
Thanks. Now it makes sense. I ll get back to you in a bit about the POSTing of the results.

**Comment by spMohanty** [comment]
@oliverroick : So I managed to push the results to GeoKey for one GeoTag-X project. But we will have multiple such projects. Can you let me know how can I easily (and possibly programmatically) create projects ?

**Comment by spMohanty** [comment]
@elliec , @oliverroick : https://github.com/geotagx/geotagx-geokey-sync
This repository handles the syncing of the data between geotagx and geokey. Currently the project creation on geokey has to be done manually and the project_id has to be manually mapped to the corresponding geotagx project name in `settings.py` included in the repository. I have also manually created the projects on geokey for all the online projects on geotagx, and synced the geojson summary data.

**Comment by oliverroick** [comment]
:+1: Did you manage to solve the issues you had yesterday? I saw the exceptions coming up in our logs.

**Comment by spMohanty** [comment]
@oliverroick : Yeah, ran into a few glitches in the beginning, in the end managed to get everything working. The code in the repository above works out of the box :D :D

**Comment by elliec** [comment]
@spMohanty can we close this?
